### PR TITLE
RzCore: save/restore bits and arch config values while printing

### LINKED ISSUE #2185: Deleting arch hint does not work
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Arch Linux
| File format of the file you reverse (mandatory)      | Any
| Architecture/bits of the file (mandatory)            | x86_64
| `rizin -v` full output, **not truncated** (mandatory)  | rizin 0.4.0-git @ linux-x86-64 commit: 4487d7e1ac8ec0346f0f0b6f14dfdc7d5e424b34, build: 2022-01-08__15:06:38

### Expected behavior
```
[0x00000000]> wx 31ed4989d1
[0x00000000]> pd 1
            0x00000000      xor   ebp, ebp
[0x00000000]> aha ppc
[0x00000000]> pd 1
            0x00000000      lbz   r10, -0x12cf(r9)
[0x00000000]> aha 0
[0x00000000]> pd 1
            0x00000000      xor   ebp, ebp
[0x00000000]> ah-*
[0x00000000]> pd 1
            0x00000000      xor   ebp, ebp
```

### Actual behavior
```
[0x00000000]> wx 31ed4989d1
[0x00000000]> pd 1
            0x00000000      xor   ebp, ebp
[0x00000000]> aha ppc
[0x00000000]> pd 1
            0x00000000      lbz   r10, -0x12cf(r9)
[0x00000000]> aha 0
[0x00000000]> pd 1
            0x00000000      lbz   r10, -0x12cf(r9)
[0x00000000]> ah-*
[0x00000000]> pd 1
            0x00000000      lbz   r10, -0x12cf(r9)
```

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

rz_analysis_op calls archbits in RzCoreBind, which sets the arch/bits according to the hints. However, when you delete a hint it finds that they are NULL, thus it does not set them anymore, leaving the older hint in place. Fix this by saving/restoring the config options before doing print_disasm.

NOTE: `pdi` does not work any way, because it uses rz_asm_op instead of rz_analysis_op, but this is not fixed here because it would be a mess...

**Test plan**

...

**Closing issues**

Closes #2185

### CODE REVIEW

**Comment by ret2libc** [comment]
cc @pelijah

**Comment by XVilka** [comment]
```
WARNING: rz_config_get: assertion 'cfg && RZ_STR_ISNOTEMPTY(name)' failed (line 76)
```

**Comment by ret2libc** [comment]
> ```
> WARNING: rz_config_get: assertion 'cfg && RZ_STR_ISNOTEMPTY(name)' failed (line 76)
> ```

I forgot the final NULL.

**Review by wargio — APPROVED** [review]

File: librz/core/disasm.c [hunk 1]
```diff
@@ -5287,6 +5287,9 @@ RZ_API int rz_core_print_disasm(RzPrint *p, RzCore *core, ut64 addr, ut8 *buf, i
 		rz_reg_arena_push(core->analysis->reg);
 	}
 
+	RzConfigHold *rch = rz_config_hold_new(core->config);
```
Comment (wargio) [inline librz/core/disasm.c]: alloc failure check?

**Comment by wargio** [comment]
image

**Comment by ret2libc** [comment]
I have removed from `ds`, in rz_print_disasm, the show_offseg variable and just got the value directly from RzCOnfig. This is because at the moment `ds` is initialized the value of asm.segoff might be wrong and it will be updated accordingly to RzAnalysisHints or whatever else could change that variable.

I think the same problem could happen for other vars as well...

File: test/db/cmd/cmd_pd [hunk 2]
```diff
@@ -2475,3 +2475,32 @@ EXPECT=<<EOF
             0x00000812      sta   0x03ff,x                             ; a,x,0x03ff,+,=[1]
 EOF
 RUN
+
+NAME=remove analysis hint
+FILE==
+CMDS=<<EOF
+e asm.functions=false
```
Comment (XVilka) [inline test/db/cmd/cmd_pd:2483]: `e asm.arch=x86`

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**